File: lib/protocol.js
'use strict';

const SOX = 0xF0;
const EOX = 0xF7;
const MANUFACTURER = [0x00, 0x21, 0x45];

// Hardware layout: 6 banks × 12 preset slots.
// These are the single source of truth for the JS side; the TUI, CLI, and
// device layer all import them from here.
const BANKS = 6;
const SLOTS_PER_BANK = 12;

const OP = {
  UPLOAD:      0x01,
  REQUEST:     0x02,
  RESPONSE:    0x01, // same byte as UPLOAD on the wire — direction disambiguates
  REMOVE:      0x05, // remove/clear files in a slot
  SWITCH_SLOT: 0x09, // "switch preset slot" — switches to AND loads the preset
  SELECT_SLOT: 0x14, // "set preset slot" — arms the target slot for upload; does NOT load it
  ACK_NACK:    0x7E,
};

const RES = {
  PRESET:  0x01,
  SLOT:    0x08, // bank/slot pair — used by select/switch/clear
  LUA:     0x0C,
  INFO:    0x7F,
};

function frame(...bytes) {
  return [SOX, ...MANUFACTURER, ...bytes, EOX];
}

const infoRequest = () => frame(OP.REQUEST, RES.INFO);

function presetRequest(bank, slot) {
  if (bank != null && slot != null) return frame(OP.REQUEST, RES.PRESET, bank, slot);
  return frame(OP.REQUEST, RES.PRESET);
}

/**
 * Build a preset-upload SysEx message.
 *
 * IMPORTANT: the device always uploads to the *currently active* slot. There
 * is no bank/slot variant of the upload command — the JSON body follows the
 * resource byte directly. To target a specific slot, send presetSlotSelect()
 * first to arm it, then upload. (Earlier code inserted bank/slot bytes here,
 * which corrupted the JSON body and got a NACK.)
 *
 * Bodies are encoded as 7-bit ASCII (SysEx data bytes must be < 0x80);
 * Node's 'ascii' encoding strips the high bit, so non-ASCII characters are
 * mangled rather than corrupting the stream.
 */
function presetUpload(jsonStr) {
  const payload = Array.from(Buffer.from(jsonStr, 'ascii'));
  return [SOX, ...MANUFACTURER, OP.UPLOAD, RES.PRESET, ...payload, EOX];
}

/** "Set preset slot" — arm the given bank/slot as the active upload target. */
function presetSlotSelect(bank, slot) {
  return frame(OP.SELECT_SLOT, RES.SLOT, bank, slot);
}

/** "Switch preset slot" — switch the device to the given bank/slot and load it. */
function presetSlotSwitch(bank, slot) {
  return frame(OP.SWITCH_SLOT, RES.SLOT, bank, slot);
}

/**
 * "Clear preset slot" — permanently removes all files in the slot (preset +
 * Lua), freeing a burned/corrupt slot. (op 0x05, resource 0x08)
 */
function clearSlot(bank, slot) {
  return frame(OP.REMOVE, RES.SLOT, bank, slot);
}

function luaRequest(bank, slot) {
  if (bank != null && slot != null) return frame(OP.REQUEST, RES.LUA, bank, slot);
  return frame(OP.REQUEST, RES.LUA);
}

/**
 * Build a Lua-upload SysEx message.
 *
 * Like presetUpload, the device always uploads to the *currently active* slot;
 * there is no bank/slot variant. The source follows the resource byte directly.
 * To target a specific slot, send presetSlotSelect() first to arm it, then
 * upload. (Inserting bank/slot bytes here corrupts the Lua body — the same bug
 * that was fixed for presetUpload.)
 */
function luaUpload(luaStr) {
  const payload = Array.from(Buffer.from(luaStr, 'ascii'));
  return [SOX, ...MANUFACTURER, OP.UPLOAD, RES.LUA, ...payload, EOX];
}
/**
 * Acknowledgement codes carried in the byte after the 0x7E op.
 *
 * Empirically, an upload triggers two 0x7E messages from the device:
 *   F0 00 21 45 7E 05 F7          ← a notification (NOT a result)
 *   F0 00 21 45 7E 01 00 00 F7    ← the actual ACK
 *
 * The old code treated any 0x7E whose code byte wasn't 0x01 as a NACK, so the
 * leading 0x05 notification was misread as a rejection. We now decode the code
 * explicitly and ignore anything that isn't a definite ack/nack.
 */
const ACK = 0x01;
const NAK = 0x00;

function classify(msg) {
  // Shortest valid message is 7 bytes: F0 00 21 45 <op> <code/res> F7.
  if (
    msg.length < 7 ||
    msg[0] !== SOX ||
    !MANUFACTURER.every((b, i) => msg[i + 1] === b) ||
    msg.at(-1) !== EOX
  ) return { type: 'unknown' };

  const op = msg[4];
  if (op === OP.RESPONSE) return { type: 'data', resource: msg[5], payload: msg.slice(6, -1) };
  if (op === OP.ACK_NACK) {
    const code = msg[5];
    if (code === ACK) return { type: 'ack' };
    if (code === NAK) return { type: 'nack' };
    return { type: 'status', code }; // e.g. 0x05 notification — wait for the real ack/nack
  }
  return { type: 'unknown' };
}

function decodeJSON(payload) {
  try {
    return JSON.parse(Buffer.from(payload).toString('ascii'));
  } catch (e) {
    throw new Error(`Malformed JSON in device response: ${e.message}`);
  }
}

function decodeText(payload) {
  return Buffer.from(payload).toString('ascii');
}

module.exports = {
  SOX,
  EOX,
  MANUFACTURER,
  BANKS,
  SLOTS_PER_BANK,
  OP,
  RES,
  frame,
  infoRequest,
  presetRequest,
  presetUpload,
  presetSlotSelect,
  presetSlotSwitch,
  clearSlot,
  luaRequest,
  luaUpload,
  classify,
  decodeJSON,
  decodeText,
};

File: mac/Sources/ElectraKit/Protocol.swift
import Foundation

/// Electra One SysEx framing and message classification.
///
/// Mirrors lib/protocol.js. Manufacturer ID is `00 21 45`. Requests use op
/// `0x02`; data dumps/uploads use op `0x01`. Uploads always go to the *active*
/// slot — to target a slot, arm it first with `presetSlotSelect` (`0x14 0x08`).
public enum E1Proto {
    public static let sox: UInt8 = 0xF0
    public static let eox: UInt8 = 0xF7
    public static let manufacturer: [UInt8] = [0x00, 0x21, 0x45]

    /// Hardware layout: 6 banks × 12 preset slots (mirrors lib/protocol.js).
    public static let banks = 6
    public static let slotsPerBank = 12

    enum Op {
        static let upload: UInt8 = 0x01
        static let request: UInt8 = 0x02
        static let response: UInt8 = 0x01
        static let remove: UInt8 = 0x05
        static let switchActive: UInt8 = 0x09   // "Switch" — activates/loads a resource
        static let selectSlot: UInt8 = 0x14     // "updateRuntime" — Set Preset Slot only arms it
        static let ackNack: UInt8 = 0x7E
    }

    enum Res {
        static let preset: UInt8 = 0x01
        static let lua: UInt8 = 0x0C
        static let info: UInt8 = 0x7F
        static let runtime: UInt8 = 0x7E
    }

    // ACK/NACK codes carried in the byte after the 0x7E op.
    static let ack: UInt8 = 0x01
    static let nak: UInt8 = 0x00

    static func frame(_ bytes: [UInt8]) -> [UInt8] {
        [sox] + manufacturer + bytes + [eox]
    }

    static func frame(_ bytes: UInt8...) -> [UInt8] {
        frame(bytes)
    }

    /// Clamp a bank into its hardware range — `UInt8(_:)` would trap on
    /// out-of-range Ints, and the API is non-throwing.
    private static func bankByte(_ bank: Int) -> UInt8 {
        UInt8(min(max(bank, 0), banks - 1))
    }

    /// Clamp a slot into its hardware range (see `bankByte`).
    private static func slotByte(_ slot: Int) -> UInt8 {
        UInt8(min(max(slot, 0), slotsPerBank - 1))
    }

    /// SysEx data bytes must be 7-bit (< 0x80), but UTF-8 emits high bytes for
    /// non-ASCII characters — which would corrupt the stream. Replace such
    /// scalars with '?'; the JS twin (lib/protocol.js) likewise coerces bodies
    /// to 7-bit via Node's 'ascii' encoding.
    private static func asciiBody(_ text: String) -> [UInt8] {
        text.unicodeScalars.map { $0.isASCII ? UInt8($0.value) : 0x3F /* '?' */ }
    }

    public static func infoRequest() -> [UInt8] {
        frame(Op.request, Res.info)
    }

    public static func presetRequest(bank: Int?, slot: Int?) -> [UInt8] {
        if let b = bank, let s = slot {
            return frame(Op.request, Res.preset, bankByte(b), slotByte(s))
        }
        return frame(Op.request, Res.preset)
    }

    public static func luaRequest(bank: Int?, slot: Int?) -> [UInt8] {
        if let b = bank, let s = slot {
            return frame(Op.request, Res.lua, bankByte(b), slotByte(s))
        }
        return frame(Op.request, Res.lua)
    }

    /// Build a preset-upload message. The JSON body follows the resource byte
    /// directly — there is no bank/slot variant of the upload command.
    public static func presetUpload(json: String) -> [UInt8] {
        frame([Op.upload, Res.preset] + asciiBody(json))
    }

    /// "Set preset slot" (op `0x14`, res `0x08`) — arm the given bank/slot as the
    /// target for subsequent file operations. Does NOT load the preset.
    public static func presetSlotSelect(bank: Int, slot: Int) -> [UInt8] {
        frame(Op.selectSlot, 0x08, bankByte(bank), slotByte(slot))
    }

    /// "Switch preset slot" (op `0x09`, res `0x08`) — make the slot active and
    /// load its preset (the controller switches to display/run it).
    public static func presetSlotSwitch(bank: Int, slot: Int) -> [UInt8] {
        frame(Op.switchActive, 0x08, bankByte(bank), slotByte(slot))
    }

    /// Upload a Lua script to the active slot (op `0x01`, resource `0x0C`).
    public static func luaUpload(source: String) -> [UInt8] {
        frame([Op.upload, Res.lua] + asciiBody(source))
    }

    /// "Clear preset slot" (op `0x05`, resource `0x08`) — permanently removes
    /// all files in the slot (preset + Lua), freeing a burned/corrupt slot.
    public static func clearSlot(bank: Int, slot: Int) -> [UInt8] {
        frame(Op.remove, 0x08, bankByte(bank), slotByte(slot))
    }

    /// A classified inbound SysEx message.
    public enum Message {
        case data(resource: UInt8, payload: [UInt8])
        case ack
        case nack
        case status(code: UInt8)   // e.g. 0x05 notification — not a result
        case unknown
    }

    public static func classify(_ msg: [UInt8]) -> Message {
        // Shortest valid message is 7 bytes: F0 00 21 45 <op> <code/res> F7.
        guard msg.count >= 7,
              msg[0] == sox,
              Array(msg[1...3]) == manufacturer,
              msg.last == eox
        else { return .unknown }

        let op = msg[4]
        if op == Op.response {
            let payload = Array(msg[6..<(msg.count - 1)])
            return .data(resource: msg[5], payload: payload)
        }
        if op == Op.ackNack {
            switch msg[5] {
            case ack: return .ack
            case nak: return .nack
            default:  return .status(code: msg[5])
            }
        }
        return .unknown
    }
}
